src/net/posix-stack.cc

```diff
@@ -897,7 +909,7 @@ future<datagram>
 posix_datagram_channel::receive() {
     _recv.prepare();
     return _fd.recvmsg(&_recv._hdr).then([this] (size_t size) {
-        socket_address dst;
+        socket_address dst = _address; // Default for when no pktinfo is available (e.g. AF_UNIX).
```

Comment:
Assuming the compiler doesn't optimize it (and it seems it shouldn't), it could be replaced by an IIFE to avoid copying in the loop

Reply:
> avoid copying in the loop

Each assignment to the `dst` in the loop is followed by `break`, so I don't really see what did you mean.

Reply:
> > avoid copying in the loop
> 
> Each assignment to the `dst` in the loop is followed by `break`, so I don't really see what did you mean.

The assignment is what I meant by copying. In that version of the code, the default constructor of `socket_address` was first invoked and then followed by the copy assignment operator. It could be reduced to just one call to the right constructor by employing some kind of IFFE, e.g.
```cpp
socket_address dst = [&] {
    for (auto* cmsg = CMSG_FIRSTHDR(&_recv._hdr); cmsg != nullptr; cmsg = CMSG_NXTHDR(&_recv._hdr, cmsg)) {
        if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_PKTINFO) {
            return ipv4_addr(copy_reinterpret_cast<in_pktinfo>(CMSG_DATA(cmsg)).ipi_addr, _address.port());
        } else if (cmsg->cmsg_level == IPPROTO_IPV6 && cmsg->cmsg_type == IPV6_PKTINFO) {
            return ipv6_addr(copy_reinterpret_cast<in6_pktinfo>(CMSG_DATA(cmsg)).ipi6_addr, _address.port());
        }
    }
    return socket_address{};
} ();
```
Here the compiler should leverage copy elision by constructing the object in place. I'm not very knowledgeable in the specifics of compilers, so maybe it's not that easy, but that was the idea I had in mind.

Reply:
It's okay, copying 128 bytes is just 16 instructions (or 8 if avx is enabled). It's not material compared to the rest of the work needed to send udp (which involves a syscall as we don't have an io_uring path for it).